### PR TITLE
return error when user write a invalid nodemask for ipam

### PR DESCRIPTION
**Pull request by konghui**
When user write a invalid nodenetmask , It will return no error. But it will return nil on call 
https://github.com/cilium/ipam/blob/2f672ef3ad54ecaef97003e66f7fd8a0608801e0/cidrset/cidr_set.go#L173

It cause this issue: https://github.com/cilium/cilium/issues/12956

I also think it better to add a error log on here to prevent return a nil value to the nodecidr list.

https://github.com/cilium/cilium/blob/758539bb0d7dec2ac4095cf4bf07ad21183fc647/pkg/ipam/allocator/podcidr/podcidr.go#L808

### CODE REVIEW

File: cidrset/cidr_set.go
Comment (qmonnet) [inline cidrset/cidr_set.go:53]:
```suggestion
	// the default subnet mask should be lower or equal to the max IPv4 netmask
```

File: cidrset/cidr_set.go
Comment (qmonnet) [inline cidrset/cidr_set.go:55]:
```suggestion
	// the default subnet mask should be lower or equal to the max IPv6 netmask
```

File: cidrset/cidr_set.go
Comment (qmonnet) [inline cidrset/cidr_set.go:71]: Better reuse the constants we've set above:
```suggestion
	// ErrSubNetMaskSizeInvalid occurs when the subnet mask size is invalid:
	// bigger than 32 for IPv4 or bigger than 128 for IPv6
	ErrSubNetMaskSizeInvalid = fmt.Errorf(
		"SubNetMask is invalid, should be lower or equal to %d for IPv4 and to %d for IPv6",
		maxSubNetMaskSizeIPv4, maxSubNetMaskSizeIPv6)
```

File: cidrset/cidr_set.go
Comment (qmonnet) [inline cidrset/cidr_set.go:87]: I would reverse the order of the checks. There's no point in checking if the difference for the subnets if we can already tell the submask provided is `>128`.
 
```suggestion
		if subNetMaskSize > maxSubNetMaskSizeIPv6 {
			return nil, ErrSubNetMaskSizeInvalid
		}
		if subNetMaskSize-clusterMaskSize > clusterSubnetMaxDiff {
			return nil, ErrCIDRSetSubNetTooBig
		}
```

File: cidrset/cidr_set_test.go
Comment (qmonnet) [inline cidrset/cidr_set_test.go:750]:
```suggestion
			description:    "Check valid subnet size with IPv4",
```

File: cidrset/cidr_set_test.go
Comment (qmonnet) [inline cidrset/cidr_set_test.go:756]:
```suggestion
			description:    "Check valid subnet mask size with IPv6",
```

File: cidrset/cidr_set_test.go
Comment (qmonnet) [inline cidrset/cidr_set_test.go:762]:
```suggestion
			description:    "Check invalid subnet size with IPv4",
```

File: cidrset/cidr_set_test.go
Comment (qmonnet) [inline cidrset/cidr_set_test.go:768]:
```suggestion
			description:    "Check invalid subnet mask size with IPv6",
```

**Review by qmonnet — CHANGES_REQUESTED** [review]
A few nits, mostly on typos, but the PR looks good overall, thanks for this contribution!

I see this is the first PR on this repo so I'm not entirely sure of the process here, I'll defer to @aanm.
I note that you don't have a `Signed-off-by:` tag in your commit log, would you please consider [adding one](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#developer-s-certificate-of-origin) when addressing the changes?

File: cidrset/cidr_set.go
Comment (konghui) [reply cidrset/cidr_set.go:53]: done

File: cidrset/cidr_set.go
Comment (konghui) [reply cidrset/cidr_set.go:55]: done

File: cidrset/cidr_set.go
Comment (konghui) [reply cidrset/cidr_set.go:71]: done

File: cidrset/cidr_set.go
Comment (konghui) [reply cidrset/cidr_set.go:87]: done

File: cidrset/cidr_set_test.go
Comment (konghui) [reply cidrset/cidr_set_test.go:756]: done

File: cidrset/cidr_set_test.go
Comment (konghui) [reply cidrset/cidr_set_test.go:762]: done

File: cidrset/cidr_set_test.go
Comment (konghui) [reply cidrset/cidr_set_test.go]: done

**Review by qmonnet — APPROVED** [review]
[I can't mark the discussions above as resolved, I guess because I don't have write access on this repo.]

Looks good, thanks a lot!

**Review by aanm — APPROVED** [review]

**Pull request merged**
